### PR TITLE
recursively look up fragments

### PR DESCRIPTION
**Pull request by minasmart**

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -73,10 +73,26 @@ function optimizeAndCompileSchema({schema, profileDocuments, compiler}) {
...
+
   return readFiles(fragmentFiles).then((fragments) => {
-    return source.concat(fragments.map((fragment) => fragment.body).join('\n'));
+    return Promise.all(fragments.map((fragment) => {
+      return recursivelySourceFragmentFiles(fragment.body, fragment.path);
+    }).concat(Promise.resolve(fragments)));
+  }).then((fragmentLists) => {
+    return fragmentLists.reduce((acc, fragments) => {
+      return acc.concat(fragments);
+    }, []);
```
Comment (minasmart) [inline src/index.js]: @swalkinshaw this is the primary change. If you can think of a better way, please let me know. This is asynchronously recursive, so it's kinda hard to follow.

File: src/index.js [hunk 2]
```diff
@@ -73,10 +73,26 @@ function optimizeAndCompileSchema({schema, profileDocuments, compiler}) {
 }
 
 function prependFragments(source, id) {
+  return recursivelySourceFragmentFiles(source, id).then((fragments) => {
+    return source.concat(fragments.map((fragment) => fragment.body).join('\n'));
+  });
+}
+
+function recursivelySourceFragmentFiles(source, id) {
```
Comment (swalkinshaw) [inline src/index.js]: Minor thing, but can this argument order match up? Just a little weird to see `source, id` then it calls `id, source`.

Comment (minasmart) [reply src/index.js]: I'm not sure which direction to go here. The rollup convention is `source, id`. In a bunch of spots in the library this thing depends on the convention is `path, body`, which translates to `id, source`. No matter which way I go, it's inconsistent somewhere. What I did here was match the rollup convention. Thoughts?

**Comment by minasmart** [comment]
@swalkinshaw I added one more thing to dedupe fragments. Can you have one more look?

File: test/fixtures/sample-project/src/graphql/query-with-duplicate-fragments.graphql [hunk 3]
```diff
@@ -0,0 +1,6 @@
+query FancyQuery($id: ID!) {
+  node(id: $id) {
+    ...ProductFragment
+    ...ProductFragment
+  }
```
Comment (minasmart) [inline test/fixtures/sample-project/src/graphql/query-with-duplicate-fragments.graphql]: This is valid graphql, and it does generate duplicate fragment declarations before the deduping logic was added, so I know the test is valid 😄
